stream/src/placeholder/utils/codec.rs

```diff
@@ -0,0 +1,141 @@
+use super::super::Error;
+use bytes::Bytes;
+use commonware_runtime::{Sink, Stream};
+
+// Sends data with a 4-byte length prefix.
+pub async fn send_frame<S: Sink>(
+    sink: &mut S,
+    buf: &[u8],
+    max_message_size: usize,
+) -> Result<(), Error> {
+    // Validate frame size
+    let n = buf.len();
+    if n == 0 {
+        return Err(Error::SendZeroSize);
+    }
+    if n > max_message_size {
+        return Err(Error::SendTooLarge(n));
+    }
+    let len: u32 = n.try_into()
+        .map_err(|_| Error::SendTooLarge(n))?;
+
+    // Send the length of the message
+    let f: [u8; 4] = len.to_be_bytes();
+    sink.send(&f)
+        .await
+        .map_err(|_| Error::SendFailed)?;
+
+    // Send the rest of the message
+    sink.send(buf)
+        .await
+        .map_err(|_| Error::SendFailed)?;
+
+    Ok(())
+}
+
+// Receives data with a 4-byte length prefix.
+pub async fn recv_frame<T: Stream>(
+    stream: &mut T,
+    max_message_size: usize,
+) -> Result<Bytes, Error> {
+    // Read the first 4 bytes to get the length of the message
+    let mut buf = [0u8; 4];
+    stream.recv(&mut buf)
+        .await
+        .map_err(|_| Error::StreamClosed)?;
+
+    // Validate frame size
+    let len = u32::from_be_bytes(buf) as usize;
+    if len > max_message_size {
+        return Err(Error::ReadTooLarge(len));
+    }
+    if len == 0 {
+        return Err(Error::StreamClosed);
+    }
+
+    // Read the rest of the message
+    let mut buf = vec![0u8; len];
```

Comment:
nit: you may be able to instantiate `Bytes` directly here and using `&mut buf` (still) without having to do the `::from()` below.

Reply:
Sorry can you write the code? The only thing I could figure out as a possibility was the following, which I don't think you meant (?)
```
    let mut buf = BytesMut::with_capacity(len);
    stream.recv(&mut buf)
        .await
        .map_err(|_| Error::ReadFailed)?;

    Ok(buf.freeze())
```

Reply:
Yeah, that approach won't work (because the `.len()` of that is 0).

I was looking around again at `Bytes::BytesMut` and think what you have now is actually probably best (disregard this). There is some overhead to interacting with `BytesMut` (over just a plain `Vec`) and what you have is probably best (and we should probably follow this approach as a rule of thumb when packing bytes everywhere going forward).